
Apply auto reload so that files are automatically loaded when a change is made


In [1]:
%reload_ext autoreload
%autoreload 2

Using Google Cloud Credentials

In [2]:
import os

from google.colab import auth
auth.authenticate_user()


In [3]:
from google.cloud import bigquery

client = bigquery.Client("tw-techdash")

Installa memory-profile for testing memory in functions

In [4]:
!pip install memory-profiler



Environment variables

In [5]:
MOUNT_POINT = os.environ['LC_MOUNT_POINT'] = "/content/drive"
SOURCE_PATH = os.environ['LC_SOURCE_PATH'] = "/content/drive/Othercomputers/My MacBook Pro/challenge_DE"


I used Google Drive to sync the files (in addition to github)

In [6]:
from google.colab import drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [7]:
drive.mount(MOUNT_POINT, force_remount=True)

target_directory = os.path.join(MOUNT_POINT, SOURCE_PATH, "src")
os.chdir(target_directory)

Mounted at /content/drive




---

### First query

In [8]:
query1= """
  WITH datos AS (
    SELECT
      FORMAT_TIMESTAMP('%Y-%m-%d', TIMESTAMP(date)) AS fecha_formateada,
      user.username as usuario
    FROM
      `tw-techdash.de_challenge_latam.tweets_latam`
  ),

  conteo_por_fecha AS (
    SELECT
      fecha_formateada,
      COUNT(usuario) AS num_tweets
    FROM
      datos
    GROUP BY
      fecha_formateada
    ORDER BY
      num_tweets DESC
    LIMIT 10
  ),

  usuario_top_por_fecha AS (
    SELECT
      fecha_formateada,
      usuario,
      COUNT(usuario) AS num_tweets_por_usuario,
      RANK() OVER (PARTITION BY fecha_formateada ORDER BY COUNT(usuario) DESC) AS rank
    FROM
      datos
    WHERE
      fecha_formateada IN (SELECT fecha_formateada FROM conteo_por_fecha)
    GROUP BY
      fecha_formateada, usuario
  )

  SELECT
    u.usuario AS top_user,
    u.fecha_formateada as date,
  FROM
    usuario_top_por_fecha u
  JOIN
    conteo_por_fecha c
  ON
    u.fecha_formateada = c.fecha_formateada
  WHERE
    u.rank = 1
  ORDER BY
    c.num_tweets DESC
"""

**import the function and variables to use:**

In [9]:
%run q1_memory.py
q1_memory

<function __main__.q1_memory(client: google.cloud.bigquery.client.Client, query: str) -> List[Tuple[str, str]]>

In [10]:
q1_memory(client, query1)


PYDEV DEBUGGER WARNING:
sys.settrace() should not be used when the debugger is being used.
This may cause the debugger to stop working correctly.
If this is needed, please check: 
http://pydev.blogspot.com/2007/06/why-cant-pydev-debugger-work-with.html
to see how to restore the debug tracing back correctly.
Call Location:
  File "/usr/local/lib/python3.10/dist-packages/memory_profiler.py", line 847, in enable
    sys.settrace(self.trace_memory_usage)


PYDEV DEBUGGER WARNING:
sys.settrace() should not be used when the debugger is being used.
This may cause the debugger to stop working correctly.
If this is needed, please check: 
http://pydev.blogspot.com/2007/06/why-cant-pydev-debugger-work-with.html
to see how to restore the debug tracing back correctly.
Call Location:
  File "/usr/local/lib/python3.10/dist-packages/memory_profiler.py", line 850, in disable
    sys.settrace(self._original_trace_function)



Filename: /content/drive/Othercomputers/My MacBook Pro/challenge_DE/src/q1_memory.py

Line #    Mem usage    Increment  Occurrences   Line Contents
     6    190.7 MiB    190.7 MiB           1   @profile
     7                                         def q1_memory(client: bigquery.Client, query: str) -> List[Tuple[str, str]]:
     8                                             """
     9                                             Execute and return the result of query and the memory
    10                                             Args:
    11                                                 client (bigquery.Client),
    12                                                 query (str)
    13                                             Returns: 
    14                                                 result of query
    15                                             """
    16    191.3 MiB      0.6 MiB           1       results = execute_query_for_memory(client, query)
    17             

[('RanbirS00614606', '2021-02-12'),
 ('MaanDee08215437', '2021-02-13'),
 ('RaaJVinderkaur', '2021-02-17'),
 ('jot__b', '2021-02-16'),
 ('rebelpacifist', '2021-02-14'),
 ('neetuanjle_nitu', '2021-02-18'),
 ('jot__b', '2021-02-15'),
 ('MangalJ23056160', '2021-02-20'),
 ('Surrypuria', '2021-02-23'),
 ('Preetm91', '2021-02-19')]

In [11]:
%run q1_time.py
q1_time

<function __main__.q1_time(client: google.cloud.bigquery.client.Client, query: str) -> Tuple[List[Tuple[str, str]], float]>

In [12]:
q1_time(client, query1)

         9102 function calls (9035 primitive calls) in 0.707 seconds

   Ordered by: cumulative time
   List reduced from 550 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      4/3    0.000    0.000    0.705    0.235 retry_unary.py:286(retry_wrapped_func)
      4/3    0.000    0.000    0.705    0.235 retry_unary.py:85(retry_target)
        2    0.000    0.000    0.705    0.353 client.py:814(_call_api)
        2    0.000    0.000    0.705    0.352 __init__.py:383(api_request)
        2    0.000    0.000    0.704    0.352 __init__.py:276(_make_request)
        2    0.000    0.000    0.704    0.352 __init__.py:345(_do_request)
        2    0.000    0.000    0.704    0.352 requests.py:482(request)
        2    0.000    0.000    0.704    0.352 sessions.py:500(request)
        2    0.000    0.000    0.696    0.348 sessions.py:673(send)
        2    0.000    0.000    0.693    0.347 adapters.py:613(send)




([('RanbirS00614606', '2021-02-12'),
  ('MaanDee08215437', '2021-02-13'),
  ('RaaJVinderkaur', '2021-02-17'),
  ('jot__b', '2021-02-16'),
  ('rebelpacifist', '2021-02-14'),
  ('neetuanjle_nitu', '2021-02-18'),
  ('jot__b', '2021-02-15'),
  ('MangalJ23056160', '2021-02-20'),
  ('Surrypuria', '2021-02-23'),
  ('Preetm91', '2021-02-19')],
 0.706737756729126)



---


### Second query:




## Note:
 Several emoji in the json appear with the unknown emoji ('\uFFFD' = �), for this reason in the WHERE clause I have decided to remove it.

In [13]:
query2 = """
  SELECT
  emoji,
  COUNT(*) AS total_emoji
FROM
  (SELECT
        REGEXP_EXTRACT_ALL(
          content,
           r"[\p{So}]"
        ) AS emojis,
        id
      FROM
        `tw-techdash.de_challenge_latam.tweets_latam`
      WHERE id is not null
  )
CROSS JOIN UNNEST(emojis) as emoji
WHERE emoji != '\uFFFD' -- �
group by emoji
order by total_emoji desc
limit 10
"""

**import the function and variables to use:**

In [14]:
%run q2_time.py
q2_time

<function __main__.q2_time(client: google.cloud.bigquery.client.Client, query: str) -> Tuple[List[Tuple[str, int]], float]>

In [15]:
q2_time(client, query2)

         8972 function calls (8905 primitive calls) in 0.760 seconds

   Ordered by: cumulative time
   List reduced from 550 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      4/3    0.000    0.000    0.759    0.253 retry_unary.py:286(retry_wrapped_func)
      4/3    0.000    0.000    0.759    0.253 retry_unary.py:85(retry_target)
        2    0.000    0.000    0.759    0.379 client.py:814(_call_api)
        2    0.000    0.000    0.759    0.379 __init__.py:383(api_request)
        2    0.000    0.000    0.758    0.379 __init__.py:276(_make_request)
        2    0.000    0.000    0.758    0.379 __init__.py:345(_do_request)
        2    0.000    0.000    0.758    0.379 requests.py:482(request)
        2    0.000    0.000    0.758    0.379 sessions.py:500(request)
        2    0.000    0.000    0.750    0.375 sessions.py:673(send)
        2    0.001    0.000    0.748    0.374 adapters.py:613(send)




([('✊', 2411),
  ('❤', 1779),
  ('☮', 316),
  ('✌', 274),
  ('♂', 184),
  ('♀', 148),
  ('✍', 103),
  ('♥', 81),
  ('⛳', 68),
  ('✅', 57)],
 0.7605245113372803)

In [16]:
%run q2_memory.py
q2_memory

<function __main__.q2_memory(client: google.cloud.bigquery.client.Client, query: str) -> List[Tuple[str, int]]>

In [17]:
q2_memory(client, query2)

Filename: /content/drive/Othercomputers/My MacBook Pro/challenge_DE/src/q2_memory.py

Line #    Mem usage    Increment  Occurrences   Line Contents
     6    192.6 MiB    192.6 MiB           1   @profile
     7                                         def q2_memory(client: bigquery.Client, query: str) ->  List[Tuple[str, int]]:
     8                                             """
     9                                             Execute and return the result of query and the memory
    10                                             Args:
    11                                                 client (bigquery.Client),
    12                                                 query (str)
    13                                             Returns: 
    14                                                 result of query
    15                                             """
    16    192.6 MiB      0.0 MiB           1       results = execute_query_for_memory(client, query)
    17            

[('✊', 2411),
 ('❤', 1779),
 ('☮', 316),
 ('✌', 274),
 ('♂', 184),
 ('♀', 148),
 ('✍', 103),
 ('♥', 81),
 ('⛳', 68),
 ('✅', 57)]

---

### Third query:

In [18]:
query3 = """
 WITH gruped AS (
  SELECT
    user.username AS username,
    m.username AS top_mentioned,
    COUNT(*) AS cuenta
  FROM
    `tw-techdash.de_challenge_latam.tweets_latam` as tweets,
    UNNEST(mentionedUsers) AS m
  WHERE tweets.id is not null
  GROUP BY
    username, top_mentioned
)

SELECT
  top_mentioned,
  SUM(cuenta) AS total_mencioned
FROM
  gruped
GROUP BY
  top_mentioned
  order by total_mencioned desc
LIMIT 10
"""

**import the function and variables to use:**

In [19]:
%run q3_time.py
q3_time

<function __main__.q3_time(client: google.cloud.bigquery.client.Client, query: str) -> Tuple[List[Tuple[str, int]], float]>

In [20]:
q3_time(client, query3)

         9204 function calls (9137 primitive calls) in 0.754 seconds

   Ordered by: cumulative time
   List reduced from 550 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      4/3    0.000    0.000    0.752    0.251 retry_unary.py:286(retry_wrapped_func)
      4/3    0.000    0.000    0.752    0.251 retry_unary.py:85(retry_target)
        2    0.000    0.000    0.752    0.376 client.py:814(_call_api)
        2    0.000    0.000    0.752    0.376 __init__.py:383(api_request)
        2    0.000    0.000    0.752    0.376 __init__.py:276(_make_request)
        2    0.000    0.000    0.752    0.376 __init__.py:345(_do_request)
        2    0.000    0.000    0.752    0.376 requests.py:482(request)
        2    0.000    0.000    0.751    0.376 sessions.py:500(request)
        2    0.000    0.000    0.748    0.374 sessions.py:673(send)
        2    0.000    0.000    0.747    0.374 adapters.py:613(send)




([('narendramodi', 2265),
  ('Kisanektamorcha', 1840),
  ('RakeshTikaitBKU', 1644),
  ('PMOIndia', 1427),
  ('RahulGandhi', 1146),
  ('GretaThunberg', 1048),
  ('RaviSinghKA', 1019),
  ('rihanna', 986),
  ('UNHumanRights', 962),
  ('meenaharris', 926)],
 0.7537240982055664)

In [21]:
%run q3_memory.py
q3_memory

<function __main__.q3_memory(client: google.cloud.bigquery.client.Client, query: str) -> List[Tuple[str, int]]>

In [22]:
q3_memory(client, query3)

Filename: /content/drive/Othercomputers/My MacBook Pro/challenge_DE/src/q3_memory.py

Line #    Mem usage    Increment  Occurrences   Line Contents
     6    192.6 MiB    192.6 MiB           1   @profile
     7                                         def q3_memory(client: bigquery.Client, query: str) ->  List[Tuple[str, int]]:
     8                                             """
     9                                             Execute and return the result of query and the memory
    10                                             Args:
    11                                                 client (bigquery.Client),
    12                                                 query (str)
    13                                             Returns: 
    14                                                 result of query
    15                                             """
    16    192.6 MiB      0.0 MiB           1       results = execute_query_for_memory(client, query)
    17            

[('narendramodi', 2265),
 ('Kisanektamorcha', 1840),
 ('RakeshTikaitBKU', 1644),
 ('PMOIndia', 1427),
 ('RahulGandhi', 1146),
 ('GretaThunberg', 1048),
 ('RaviSinghKA', 1019),
 ('rihanna', 986),
 ('UNHumanRights', 962),
 ('meenaharris', 926)]



---

# Result:

## First Query:
"Las top 10 fechas donde hay más tweets. Mencionar el usuario (username) que más publicaciones tiene por cada uno de esos días"

| top_user          | date       |
|-------------------|------------|
| RanbirS00614606   | 2021-02-12 |
| MaanDee08215437   | 2021-02-13 |
| RaaJVinderkaur    | 2021-02-17 |
| jot__b            | 2021-02-16 |
| rebelpacifist     | 2021-02-14 |
| neetuanjle_nitu   | 2021-02-18 |
| jot__b            | 2021-02-15 |
| MangalJ23056160   | 2021-02-20 |
| Surrypuria        | 2021-02-23 |
| Preetm91          | 2021-02-19 |


## Second Query:

| emoji | total_emoji |
|-------|-------|
| ✊    | 2411  |
| ❤    | 1779  |
| ☮    | 316   |
| ✌    | 274   |
| ♂    | 184   |
| ♀    | 148   |
| ✍    | 103   |
| ♥    | 81    |
| ⛳    | 68    |
| ✅    | 57    |


## Third Query:

| top_mentioned            | total_mentioned |
|---------------------|-------|
| narendramodi        | 2265  |
| Kisanektamorcha     | 1840  |
| RakeshTikaitBKU     | 1644  |
| PMOIndia            | 1427  |
| RahulGandhi         | 1146  |
| GretaThunberg       | 1048  |
| RaviSinghKA         | 1019  |
| rihanna             | 986   |
| UNHumanRights       | 962   |
| meenaharris         | 926   |



---
## Oportunidades de mejoras:

Como oportunidades de mejoras es editar el manejador de errores de una manera más específica, en la cual pueda capturar excepciones más distintivas como las BadRequest.

Otra cosa a mejorar es investigar una libería mejor para el uso de memoria, debido a la recomendada, no me dejo del todo conforme con la salida.

Otro punto más pequeño es no sincronizar con GoogleDrive el Google Colab ya que al cambiar de ambientes se desincroniza y hay que volver a actualizarlo.
